calyx-backend/src/firrtl.rs

```diff
@@ -0,0 +1,181 @@
+//! FIRRTL backend for the Calyx compiler.
+//!
+//! Transforms an [`ir::Context`](crate::ir::Context) into a formatted string that represents a
+//! valid FIRRTL program.
+
+use crate::{traits::Backend, VerilogBackend};
+use calyx_ir::{self as ir};
+use calyx_utils::{CalyxResult, Error, OutputFile};
+use std::io;
+use std::time::Instant;
+
+pub(super) const SPACING: &str = "    ";
+
+/// Implements a simple FIRRTL backend. The backend only accepts Calyx programs with no control
+/// and no groups.
+#[derive(Default)]
+pub struct FirrtlBackend;
+
+impl Backend for FirrtlBackend {
+    fn name(&self) -> &'static str {
+        "firrtl"
+    }
+
+    fn link_externs(
+        _prog: &calyx_ir::Context,
+        _write: &mut calyx_utils::OutputFile,
+    ) -> calyx_utils::CalyxResult<()> {
+        Ok(()) // FIXME: Need to implement
+    }
+
+    fn validate(prog: &calyx_ir::Context) -> calyx_utils::CalyxResult<()> {
+        VerilogBackend::validate(prog) // FIXME: would this work if we wanted to check for the same things?
+    }
+
+    fn emit(ctx: &ir::Context, file: &mut OutputFile) -> CalyxResult<()> {
+        let out = &mut file.get_write();
+        let comps = ctx.components.iter().try_for_each(|comp| {
+            // Time the generation of the component.
+            let time = Instant::now();
+            let out = emit_component(comp, out);
+            log::info!("Generated `{}` in {:?}", comp.name, time.elapsed());
+            out
+        });
+        comps.map_err(|err| {
+            let std::io::Error { .. } = err;
+            Error::write_error(format!(
+                "File not found: {}",
+                file.as_path_string()
+            ))
+        })
```

Comment:
So the above and this block could be simplified into the following:
```suggestion
    for comp in ctx.components.iter() {
        emit_component(comp, out)?
    }
```
Alternatively, if you want to preserve the above structure, you could do

```suggestion
    let out = emit_component(comp, out)?;
```

But since the return value on success is the empty tuple `()` aka "unit", there
isn't a need to explicitly return it beyond making sure that we are handling the
errors which can be done via the `?` syntax and the `Try` trait. Since the
`CalyxResult` can be built from the `std::io::Error` type (via a `From`
implementation) the `?` automatically handles the conversion between the two
types, freeing you from writing this error translation manually.

Reply:
I think this makes sense! But I think I may need to quickly talk this over the next time we meet :sweat_smile: